test/jdk/java/net/httpclient/SendResponseHeadersTest.java

```diff
@@ -0,0 +1,104 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8253005
+ * @summary Check that sendResponseHeaders throws an IOException when headers
+ *  have already been sent
+ * @run testng/othervm SendResponseHeadersTest
+ */
+
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+import com.sun.net.httpserver.HttpServer;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static org.testng.Assert.expectThrows;
+import static org.testng.Assert.fail;
+
+
+public class SendResponseHeadersTest {
+
+    @Test
+    public void testSend() throws Exception {
+        var addr = new InetSocketAddress(InetAddress.getLoopbackAddress(), 0);
+        var server = HttpServer.create(addr, 0);
+        var path = "/test/foo.html";
+        server.createContext("/test", new TestHandler());
+        ExecutorService executor = Executors.newCachedThreadPool();
+        server.setExecutor (executor);
+        server.start();
+
+        URI uri = new URI("http", null,
+                server.getAddress().getHostString(),
+                server.getAddress().getPort(),
+                path, null, null);
+
+        try {
+            HttpClient client = HttpClient.newHttpClient();
+            HttpRequest request = HttpRequest.newBuilder(uri)
+                    .GET()
+                    .build();
+            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+            // verify empty response received, otherwise an error has occurred
+            if (!response.body().isEmpty())
+                fail(response.body());
+        } finally {
+            server.stop(2);
+            executor.shutdown();
+        }
+    }
+
+    static class TestHandler implements HttpHandler {
+        public void handle(HttpExchange exchange) throws IOException {
+            try (InputStream is = exchange.getRequestBody();
+                 OutputStream os = exchange.getResponseBody()) {
+
+                is.readAllBytes();
+                exchange.sendResponseHeaders(200, 0);
+                try {
+                    IOException io = expectThrows(IOException.class,
+                            () -> exchange.sendResponseHeaders(200, "failMsg".getBytes().length));
+                    System.out.println("Got expected exception: " + io);
+                } catch (Throwable t) {
+                    // unexpected exception thrown, return error to client
+                    t.printStackTrace();
+                    os.write(("Unexpected error: " + t).getBytes());
```

Comment:
Hmm... I'd be tempted to drop this catch block altogether, it's not possible to know exactly what to do there? given that we don't know what the failure is.

Reply:
This is executed on the server side - the Throwable here is the AssertionError thrown by expectThrows if the expected exception is not thrown. It is important to return this to the client side so that the client side (main thread) can fail properly.

Reply:
Oh, ok. So the whole purpose of the catch block is to gracefully terminate things (and trigger error reporting) if the test fails - which it should never do ;-) Ok, thanks.   [ It's almost like the code would be more easily understood if written without expectThrows, in this particular case ]

Reply:
>                      // unexpected exception thrown, return error to client 

The comment is a bit misleading. Maybe it should say:

```
                     // expectThrows triggered an assertion, return error message to the client.
```

Reply:
Comment changed as requested. See https://github.com/openjdk/jdk/pull/1014/commits/8ff6d9b4fcbc0f57f136ff0e7df0d371664badb2